### PR TITLE
Re-instate the plugin BUILD

### PR DESCRIPTION
**Pull request by DavidNorman**
This change re-instates the plugin BUILD link which was removed in commit 7de939bb74c5edbc2f45e77a5d4696e70bb59e5b

### CODE REVIEW

**Comment by drpngx** [comment]
@kayzhu what do you think? This was introduced in 169160056.

**Comment by kayzhu** [comment]
@DavidNorman

Hi David, sorry for removing this in my earlier change: I wasn't quite sure what the empty BUILD file was for, and whether / how you were using it. 

I wonder:
1. depending on your need, is there an alternative solution than introducing a new directory with an empty BUILD file?
2. if not, I would suggest adding more comments in the BUILD file, or add README.md to this directory itself to indicate how it is being used, and why it is unwise to remove it :)

Thanks!

**Comment by DavidNorman** [comment]
Thanks.  I realize it looks a bit like a nothing.  However, since the repo keepers are (rightly) suggesting that our 3rd party code remains out of the main repo - I think it is important to have a sensible and clean way of integrating it.  Having to make edits to other BUILD files, where there are many more targets, seems messy.

It is a shame that there isn't a 'find and include a target if it exists' option in bazel.   I played around with a few possibilities, but as you can see in the conversations in the my original pull requests, they were considered more messy than this way.

I'll add a README.md file to that directory.

**Comment by drpngx** [comment]
I'm not 100% sure that this is needed. Can we iterate internally with @jart or @gunan and @yifeif on this?

**Comment by kayzhu** [comment]
Adding @hawkinsp as well who may have some idea too.

@drpngx Sounds good -- it's possible that they may be aware of a more standard way for this.

@DavidNorman thanks for the update! I will see if @jart @gunan @yifeif has any suggestions for a possibly better solution for this issue.

**Comment by jart** [comment]
I have limited familiarity with XLA but I understand the problem you're describing.

If they were all first party, you could have a `config_setting` like we do for jemalloc where the user passes `--config=with_xla_foo=true` to Bazel, which then gets triggered by `select()`. But if the labels are unknowable, then Bazel's design is you can only really do it by having the user define a new rule at the tip of the build tree.

For example, if XLA compiler was a cc_binary, you'd have your own rule with the standard set of plugins and a shell main.cc file that statically registers them and invokes the compiler. But since I'm assuming XLA is baked into TensorFlow's big .so file, it's probably not as trivial as writing a cc_binary() to define a custom tip for the entire TensorFlow build.

In that case, have you considered doing what custom ops do? Plugin is an alwayslink=1 library with that C++11 lambda singleton registration hack, which a Python file dynamically links at runtime.

**Comment by DavidNorman** [comment]
@jart - I've tried the dynamic linking thing but both bazel and protobuf are not up to the job.  protobuf doesn't like having multiple copies of itself in the same address space, and bazel didn't have the features for a DSO which is linked down to other DSOs.  or something like that.  i was in touch with a person from the bazel team who was fixing this, so it probably does now.

**Comment by gunan** [comment]
I think a recent change by @allenlavoie together with some pending work will help with having multiple copies of protobuf. Ill let Allen comment more on dynamic linking.

**Comment by allenlavoie** [comment]
In principle we could do the same thing for XLA extensions that we do for op libraries and filesystems: include the registries and symbols needed to register new things in //tensorflow:libtensorflow_framework.so, then use an equivalent of tf.load_op_library to load a shared object (that object references symbols in libtensorflow_framework.so and does the registration; we could also add an XLA dlopen() wrapper to the C API for access in other languages). Right now all of the XLA symbols get added to pywrap_tensorflow directly, so that would require some rearranging.

After the RTLD_GLOBAL removal you should need to explicitly include protobufs to break things; otherwise if you use the @protobuf_archive//:protobuf_headers you'll just pick up the version in libtensorflow_framework.so, which should work fine.

This sort of registration via dynamic linking would be a bit of work, so in the meantime I don't see a problem with doing plugins this way...

**Comment by DavidNorman** [comment]
That's good news regarding protobuf.  I did try, for a week, to generate a plugin DSO (hence the name of the driver directory).  I can see that protobuf was just not designed to support dynamic loading. Why would it be given where it has come from.  It doesn't have 16 bit type support for the same reason.

I would be delighted to be able to release just a loadable driver though, so I will watch the neatening of the interfaces with interest. 

cheers for the responses.

**Comment by DavidNorman** [comment]
Hi all - so until the DSO changes that are required to get proper plugins working - can we reinstate these changes?

thanks

**Comment by allenlavoie** [comment]
To be clear, I'm in favor but don't have much to do with XLA development. @kayzhu do you have objections?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by kayzhu** [comment]
@allenlavoie No objections from me. I will take another look at it shortly for final approve if other folks are fine with this change too.

**Comment by gunan** [comment]
Also no objections from me.

**Comment by gunan** [comment]
And the failures are unrelated.

**Comment by drpngx** [comment]
I'm good with that. Sorry to delay this, I thought there would be a simpler solution.

**Review by kayzhu — APPROVED** [review]

**Comment by DavidNorman** [comment]
awesome.  thanks guys

**Pull request merged**
